rust/tw_internet_computer/src/protocol/principal.rs

```diff
@@ -0,0 +1,260 @@
+///
+/// Principals are generic identifiers for canisters, users
+/// and possibly other concepts in the future.
+/// As far as most uses of the IC are concerned they are
+/// opaque binary blobs with a length between 0 and 29 bytes,
+/// and there is intentionally no mechanism to tell canister ids and user ids apart.
+///
+/// Note a principal is not necessarily tied with a public key-pair,
+/// yet we need at least a key-pair of a related principal to sign
+/// requests.
+///
+/// A Principal can be serialized to a byte array ([`Vec<u8>`]) or a text
+/// representation, but the inner structure of the byte representation
+/// is kept private.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct Principal {
+    /// Length.
+    len: u8,
+
+    /// The content buffer. When returning slices this should always be sized according to
+    /// `len`.
+    bytes: [u8; Self::MAX_LENGTH_IN_BYTES],
+}
+
+impl Principal {
+    const MAX_LENGTH_IN_BYTES: usize = 29;
+    const CRC_LENGTH_IN_BYTES: usize = 4;
+
+    const SELF_AUTHENTICATING_TAG: u8 = 2;
+    const ANONYMOUS_TAG: u8 = 4;
+
+    /// Construct a [`Principal`] of the IC management canister
+    pub const fn management_canister() -> Self {
+        Self {
+            len: 0,
+            bytes: [0; Self::MAX_LENGTH_IN_BYTES],
+        }
+    }
+
+    /// Construct a self-authenticating ID from public key
+    pub fn self_authenticating<P: AsRef<[u8]>>(public_key: P) -> Self {
+        let public_key = public_key.as_ref();
+        let hash = sha224(public_key);
+        let mut bytes = [0; Self::MAX_LENGTH_IN_BYTES];
+        bytes[..Self::MAX_LENGTH_IN_BYTES - 1].copy_from_slice(hash.as_slice());
+        bytes[Self::MAX_LENGTH_IN_BYTES - 1] = Self::SELF_AUTHENTICATING_TAG;
+
+        Self {
+            len: Self::MAX_LENGTH_IN_BYTES as u8,
+            bytes,
+        }
+    }
+
+    /// Construct an anonymous ID.
+    pub const fn anonymous() -> Self {
+        let mut bytes = [0; Self::MAX_LENGTH_IN_BYTES];
+        bytes[0] = Self::ANONYMOUS_TAG;
+        Self { len: 1, bytes }
+    }
+
+    /// Construct a [`Principal`] from a slice of bytes.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the slice is longer than 29 bytes.
+    pub const fn from_slice(slice: &[u8]) -> Self {
+        match Self::try_from_slice(slice) {
+            Ok(v) => v,
+            _ => panic!("slice length exceeds capacity"),
+        }
+    }
+
+    /// Construct a [`Principal`] from a slice of bytes.
+    pub const fn try_from_slice(slice: &[u8]) -> Result<Self, PrincipalError> {
+        const MAX_LENGTH_IN_BYTES: usize = Principal::MAX_LENGTH_IN_BYTES;
+        match slice.len() {
+            len @ 0..=MAX_LENGTH_IN_BYTES => {
+                let mut bytes = [0; MAX_LENGTH_IN_BYTES];
+                let mut i = 0;
+                while i < len {
+                    bytes[i] = slice[i];
+                    i += 1;
+                }
```

Comment:
Do I understand correctly that it can be simplified as follow?
```rust
let mut bytes = [0; MAX_LENGTH_IN_BYTES];
bytes[0..slice.len()].copy_from_slice(slice);
```

Reply:
`copy_from_slice` cannot be called in a `const` function. I checked over the code however and this function is never used in that context, so I will change the code to use this instead.

Reply:
Done